flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/compact/AggregateMergeFunction.java

```diff
@@ -0,0 +1,165 @@
+            Map<String, AggregationKind> aggregationKindMap) {
+        this.primaryKeyType = primaryKeyType;
+        this.rowType = rowType;
+        this.aggregationKindMap = aggregationKindMap;
+
+        List<LogicalType> fieldTypes = rowType.getChildren();
+        this.getters = new RowData.FieldGetter[fieldTypes.size()];
+        for (int i = 0; i < fieldTypes.size(); i++) {
+            getters[i] = RowData.createFieldGetter(fieldTypes.get(i), i);
+        }
+
+        this.isPrimaryKey = new boolean[this.getters.length];
+        List<String> rowNames = rowType.getFieldNames();
+        for (String primaryKeyName : primaryKeyType.getFieldNames()) {
+            isPrimaryKey[rowNames.indexOf(primaryKeyName)] = true;
+        }
+
+        this.aggregateFunctions = new ArrayList<>(rowType.getFieldCount());
+        for (int i = 0; i < rowType.getFieldCount(); i++) {
+            ColumnAggregateFunction<?> f = null;
+            if (aggregationKindMap.containsKey(rowNames.get(i))) {
+                f =
+                        ColumnAggregateFunctionFactory.getColumnAggregateFunction(
+                                aggregationKindMap.get(rowNames.get(i)), rowType.getTypeAt(i));
+            } else {
+                if (!isPrimaryKey[i]) {
+                    throw new IllegalArgumentException(
+                            "should  set aggregate function for every column not part of primary key");
+                }
+            }
+            aggregateFunctions.add(f);
+        }
+    }
+
+    @Override
+    public void reset() {
+        this.row = new GenericRowData(getters.length);
+    }
+
+    @Override
+    public void add(RowData value) {
+        for (int i = 0; i < getters.length; i++) {
+            Object currentField = getters[i].getFieldOrNull(value);
+            ColumnAggregateFunction<?> f = aggregateFunctions.get(i);
+            if (isPrimaryKey[i]) {
+                // primary key
+                if (currentField != null) {
+                    row.setField(i, currentField);
+                }
+            } else {
+                if (f != null) {
+                    f.reset();
+                    Object oldValue = row.getField(i);
+                    if (oldValue != null) {
+                        f.aggregate(oldValue);
+                    }
+                    switch (value.getRowKind()) {
+                        case INSERT:
+                            f.aggregate(currentField);
+                            break;
+                        case DELETE:
+                        case UPDATE_AFTER:
+                        case UPDATE_BEFORE:
+                        default:
+                            throw new UnsupportedOperationException(
+                                    "Unsupported row kind: " + row.getRowKind());
+                    }
+                    Object result = f.getResult();
+                    if (result != null) {
+                        row.setField(i, result);
```

Comment:
Value of `row` is inspected only when `AggregateMergeFunction#getValue` is called. You've stored aggregated results in `f`, why resetting it every time and aggregate twice?

My suggestion:
* In `AggregateMergeFunction#reset`, reset every aggregate function.
* In `AggregateMergeFunction#add`, aggregate each field into the corresponding function. Current results are now stored in the function.
* In `AggregateMergeFunction#getValue`, move aggregated results from functions into the row.

Reply:
> 的值`row`仅在被调用时才被检查`AggregateMergeFunction#getValue`。您已将汇总结果存储在 中`f`，为什么每次都重置它并汇总两次？
> 
> 我的建议：
> 
> * 在`AggregateMergeFunction#reset`中，重置每个聚合函数。
> * 在`AggregateMergeFunction#add`中，将每个字段聚合到相应的函数中。当前结果现在存储在函数中。
> * 在`AggregateMergeFunction#getValue`中，将聚合结果从函数移到行中。

My aggregation function f belongs to each column. If the result is not stored in row and reset, the aggregation result of another primary key on the same column will be wrong the next time I add it.
Can your method solve this problem?

I don't know when AggregateMergeFunction#reset and AggregateMergeFunction#getValue are called, and the breakpoint is not called when passing the test. Let me know if I'm sure it's okay to do this and I'll start coding

Reply:
`AggregateMergeFunction#reset` will be called before adding the first row of a new primary key, and `AggregateMergeFunction#getValue` will be called after the last row of that primary key. See `SortMergeReader` if you're intersted.

> I don't know when AggregateMergeFunction#reset and AggregateMergeFunction#getValue are called, and the breakpoint is not called when passing the test.

Are you sure about this? My breakpoint at `AggregateMergeFunction#reset` and `AggregateMergeFunction#getValue` can be triggered when running `AggregationITCase`.

Reply:
> `AggregateMergeFunction#reset` will be called before adding the first row of a new primary key, and `AggregateMergeFunction#getValue` will be called after the last row of that primary key. See `SortMergeReader` if you're intersted.
> 
> > I don't know when AggregateMergeFunction#reset and AggregateMergeFunction#getValue are called, and the breakpoint is not called when passing the test.
> 
> Are you sure about this? My breakpoint at `AggregateMergeFunction#reset` and `AggregateMergeFunction#getValue` can be triggered when running `AggregationITCase`.

I'm sorry , i got it wrong